src/persist-types/src/stats.rs

```diff
@@ -518,25 +518,37 @@ pub fn trim_to_budget(
         return;
     }
 
-    // That wasn't enough. Sort the columns in order of ascending size and
-    // keep however many fit within the budget. This strategy both keeps the
-    // largest total number of columns and also optimizes for the sort of
-    // columns we expect to need stats in practice (timestamps are numbers
-    // or small strings).
+    // That wasn't enough. Try recursively trimming out entire cols.
+    //
+    // TODO: There are certainly more elegant things we could do here. One idea
+    // would be to call `trim_to_budget_struct` but with a closure for
+    // force_keep_col that always returns false. That would potentially at least
+    // keep _something_. Another possibility would be to replace this whole bit
+    // with some sort of recursive max-cost search with force_keep_col things
+    // weighted after everything else.
+    let mut budget_shortfall = stats.encoded_len().saturating_sub(budget);
+    trim_to_budget_struct(stats, &mut budget_shortfall, &force_keep_col);
+}
```

Comment:
Ben, I tried adding the bit we discussed to throw the entire thing out if just the force keep columns put us over budget, but it broke some tests and I didn't want to spend too much time investigating. The rustdoc of this method does note that we allow going over budget in that case, so it seems Past Us decided that was fine. One thing I remembered after we talked is that the "lossy trimming" above does truncate varlen fields to a hard max, so we're insulated from badness of a single large value.